jupyterhub/tests/test_spawner.py

```diff
@@ -382,7 +382,7 @@ async def test_spawner_delete_server(app):
     assert spawner.server is None
 
 
-@pytest.mark.parametrize("name", ["has@x", "has~x", "has%x", "has%40x"])
+@pytest.mark.parametrize("name", ["has@x", "has%x", "has%40x"])
```

Comment:
This test still needs to pass. If the test is failing, JupyterHub's *behavior* needs to change, not just skipping a test.

However, [this run](https://github.com/minrk/jupyterhub/runs/1378629601) suggests that no change is needed.

```suggestion
@pytest.mark.parametrize("name", ["has@x", "has~x", "has%x", "has%40x"])
```

If there's a change in behavior across Python versions, then we are likely missing some important test matrix entries.

Reply:
I dug into it, and you're right, this test fails on Python 3.7+ with subdomains enabled. This is a bug in JupyterHub! We need to make sure that we are testing with subdomains on more versions of Python. The test matrix is too bare if it misses this, as it lacks some key combination (e.g. 3.9 + subdomain).

The issue is not in the socket library, but rather [urllib.parse.quote](https://docs.python.org/3/library/urllib.parse.html?highlight=urllib%20quote#urllib.parse.quote) which added `~` to the list of safe characters in 3.7. We need to handle this in jupyterhub, where we are using url-quoting for a domain name, which isn't quite right.

Short term answer: leave this test alone (this PR will not fail) and we can look at a new PR to add a matrix entry to cover it, and fix the bug